Add github workflow for building and publishing multi-arch image

adriananeci opened the pull request:
Resolves https://github.com/argoproj-labs/argocd-extension-installer/issues/4

jessesuen commented:
Discussed offline with @rbreeze. He can help to shepherd this in and get an official multi-arch image available.

rbreeze commented on .github/workflows/publish-image.yml:
```diff
@@ -0,0 +1,49 @@
+    steps:
+
+    - name: Checkout repository
+      uses: actions/checkout@v4
```
Please use hashes instead of SemVers for all GitHub actions references. Otherwise, this looks great!

E.g.
```
uses: actions/checkout@dslkjfn234nb1123 //v4
```

adriananeci replied:
Done

cgriego commented:
Will there be a new release when this is merged? I ask because the current on has high severity vulnerabilities.
https://quay.io/repository/argoprojlabs/argocd-extension-installer/manifest/sha256:f50fa11a4592f3fcdd5a137dab8ed32067bb779a77a393f179e8a5d96abe1a80?tab=vulnerabilities

SantiEich commented:
Hi! I need this too, can someone please merge this? Thanks

rbreeze reviewed (approved):
LGTM

chit786 commented:
+1

hungbbear commented:
HI, Can we merge this PR now?

leoluz commented on .github/workflows/publish-image.yml:
```diff
@@ -0,0 +1,49 @@
+    - name: Log in to the Container registry
+      uses: docker/login-action@343f7c4344506bcbf9b4de18042ae17996df046d # v3
+      with:
+        registry: ${{ env.REGISTRY }}
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
```
Not sure what is the intention here. Should we configure those additional repository secrets? We already have secrets to push the images to quay.io. I think that we should use those instead.

adriananeci replied:
I'm don't have access to see those secrets so I preferred to use the ghcr.io registry and the `GITHUB_TOKEN` which is automatically created at the start of each workflow job.

leoluz replied:
We had problems in the past publishing images in multiple registries. We officially use quay.io to publish Argo related images now. Images for this repo are published here:
https://quay.io/repository/argoprojlabs/argocd-extension-installer?tab=tags
You can use the variables `QUAY_USERNAME` and `QUAY_TOKEN` for authenticating.

adriananeci replied:
done, switched to the quay related envs

leoluz commented on .github/workflows/publish-image.yml:
```diff
@@ -0,0 +1,49 @@
+
+on:
+  push:
+    branches: [ "main" ]
```
Not sure if we should configure it to  push a new image on every merge to master. I think a better approach would be to run the job when a new tag is pushed.

adriananeci replied:
I'm fine either way, my intention was to have a latest image version bound to the head of the main branch. But I guess this can be adjusted in a followup PR once a decision is made in regards to the triggers

leoluz replied:
Let's just trigger the release on push tags events:
```
on:
  push:
    tags:
      - 'v*'
```

adriananeci replied:
done

leoluz commented on .github/workflows/publish-image.yml:
```diff
@@ -0,0 +1,49 @@
+      with:
+        images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@68827325e0b33c7199eb31dd4e31fbe9023e06e3 # v3
```
Can you please clarify why QEMU is necessary in this job?

adriananeci replied:
For multi-arch support

leoluz commented on .github/workflows/publish-image.yml:
```diff
@@ -0,0 +1,49 @@
+        context: .
+        platforms: linux/amd64,linux/arm64,linux/arm/v6,linux/arm/v7,linux/arm64/v8
+        push: true
+        tags: ${{ steps.meta.outputs.tags }}
```
What exactly would be the tag in this case? It seems that what we need instead is to add the git tag.

adriananeci replied:
See https://github.com/docker/metadata-action?tab=readme-ov-file#usage for more details. It depends on the github events that is triggering the workflow. The intend here is to react on the push tag and pust to the main branch events for releasing new versions and the `latest` one

leoluz replied:
Thank you. I haven't used `build-push-action` action before but it looks useful.
According to the docs we are going to have:
```
push		refs/heads/main		main
push tag	refs/tags/v1.2.3	v1.2.3, latest
```
Let's keep just the `push tag` event triggering the release to avoid new images on unrelated commits like docs for example.

adriananeci replied:
done, adjusted it to trigger the release on push tags events only.

leoluz reviewed (changes requested):
Please check my comments

PatrickLaflamme commented:
+1, I see @adriananeci responded to your comments, @leoluz. Are you comfortable with his responses? I'd love to get multi-arch support asap as this is a blocker for my team.

leoluz commented on .github/workflows/publish-image.yml:
```diff
@@ -0,0 +1,49 @@
+    branches: [ "main" ]
+
+env:
+  REGISTRY: ghcr.io
```
It should be `quay.io`

leoluz commented on .github/workflows/publish-image.yml:
```diff
@@ -0,0 +1,49 @@
+
+env:
+  REGISTRY: ghcr.io
+  IMAGE_NAME: ${{ github.repository }}
```
The image URL should resolve with the following URI: `quay.io/argoprojlabs/argocd-extension-installer:<tag>`
We need to add the prefix `argoprojlabs/` in the image name.

leoluz reviewed (changes requested):
Thank you @adriananeci for working on this. 
I added a few more comments. Please take a look.

leoluz reviewed (changes requested):
Thank you @adriananeci for addressing the comments. This PR LGTM now.

The only missing piece is fixing the DCO PR check. You must sign off your commits before I am able to merge. There are detailed instructions in the job showing you the Git commands to execute: https://github.com/argoproj-labs/argocd-extension-installer/pull/5/checks?check_run_id=23539047093

adriananeci commented:
> Thank you @adriananeci for addressing the comments. This PR LGTM now.
> 
> The only missing piece is fixing the DCO PR check. You must sign off your commits before I am able to merge. There are detailed instructions in the job showing you the Git commands to execute: https://github.com/argoproj-labs/argocd-extension-installer/pull/5/checks?check_run_id=23539047093

Done

leoluz reviewed (approved):
LGTM.
Thank you!
